File: libs/java/jmstest/src/test/java/com/broker/itest/jms/selector/SelectorTest.java
/*
 * Copyright 2014-present IVK JSC. All Rights Reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
 
package com.broker.itest.jms.selector;

import static org.junit.Assert.assertEquals;
import static org.junit.Assert.assertTrue;
import static org.junit.Assert.fail;

import javax.jms.DeliveryMode;
import javax.jms.JMSException;
import javax.jms.TextMessage;

import org.junit.Rule;
import org.junit.Test;
import org.junit.rules.TestName;

import com.broker.itest.provider.ProviderLoader;
import com.broker.itest.testcase.QueueConnectionTestCase;

/**
 * Test the message selector features of JMS
 */
public class SelectorTest extends QueueConnectionTestCase {

	@Rule
	public TestName name = new TestName();

	@Override
	public String getTestName() {
		return name.getMethodName();
	}

	/**
	 * Test that an empty string as a message selector indicates that there is no message selector for the message
	 * consumer.
	 */
	@Test
	public void testEmptyStringAsSelector() {
		try {
			//need close previous receiver for this session
			receiver.close();
			receiverConnection.stop();

			receiver = receiverSession.createReceiver(receiverQueue, "");
			receiverConnection.start();

			TextMessage message = senderSession.createTextMessage();
			message.setText("testEmptyStringAsSelector");
			sender.send(message);

			TextMessage msg = (TextMessage) receiver.receive(ProviderLoader.TIMEOUT);
			assertTrue("No message was received", msg != null);
			if (msg == null)
				return;
			assertEquals("testEmptyStringAsSelector", msg.getText());
		} catch (JMSException e) {
			fail(e.getMessage());
		}
	}

	/**
	 * Tats that String literals are well handled by the message selector. <br />
	 * <ul>
	 * <li><code>"string = 'literal''s;"</code> is <code>true</code> for "literal's" and <code>false</code> for
	 * "literal"</li>
	 * </ul>
	 */

	@Test
	public void testStringLiterals() {
		try {
			//need close previous receiver for this session
			receiver.close();
			receiverConnection.stop();

			receiver = receiverSession.createReceiver(receiverQueue, "string = 'literal''s'");
			receiverConnection.start();

			TextMessage dummyMessage = senderSession.createTextMessage();
			dummyMessage.setStringProperty("string", "literal");
			dummyMessage.setText("testStringLiterals:1");
			sender.send(dummyMessage);

			TextMessage message = senderSession.createTextMessage();
			message.setStringProperty("string", "literal's");
			message.setText("testStringLiterals:2");
			sender.send(message);

			TextMessage msg = (TextMessage) receiver.receive(ProviderLoader.TIMEOUT);
			assertTrue("No message was received", msg != null);
			if (msg == null)
				return;
			assertEquals("testStringLiterals:2", msg.getText());
		} catch (JMSException e) {
			fail(e.getMessage());
		}
	}

	/**
	 * Test that the JMS property <code>JMSDeliveryMode</code> is treated as having the values <code>'PERSISTENT'</code>
	 * or <code>'NON_PERSISTENT'</code> when used in a message selector (chapter 3.8.1.3).
	 */
	@Test
	public void testJMSDeliveryModeInSelector() {
		try {
			//need close previous receiver for this session
			receiver.close();
			receiverConnection.stop();

			receiver = receiverSession.createReceiver(receiverQueue, "JMSDeliveryMode = 'PERSISTENT'");
			receiverConnection.start();

			TextMessage dummyMessage = senderSession.createTextMessage();
			dummyMessage.setText("testJMSDeliveryModeInSelector:1");
			// send a dummy message in *non persistent* mode
			sender.send(dummyMessage, DeliveryMode.NON_PERSISTENT, sender.getPriority(), sender.getTimeToLive());

			TextMessage message = senderSession.createTextMessage();
			message.setText("testJMSDeliveryModeInSelector:2");
			// send a message in *persistent*
			sender.send(message, DeliveryMode.PERSISTENT, sender.getPriority(), sender.getTimeToLive());

			TextMessage msg = (TextMessage) receiver.receive(ProviderLoader.TIMEOUT);
			assertTrue("No message was received", msg != null);
			if (msg == null)
				return;
			// only the message sent in persistent mode should be received.
			assertEquals(DeliveryMode.PERSISTENT, msg.getJMSDeliveryMode());
			assertEquals("testJMSDeliveryModeInSelector:2", msg.getText());
		} catch (JMSException e) {
			fail(e.getMessage());
		}
	}

	@Test
	/**
	 * Test that conversions that apply to the <code>get</code> methods for properties do not apply when a property is
	 * used in a message selector expression. Based on the example of chapter 3.8.1.1 about identifiers.
	 */
	public void testIdentifierConversion() {
		try {
			//need close previous receiver for this session
			receiver.close();
			receiverConnection.stop();

			receiver = receiverSession.createReceiver(receiverQueue, "NumberOfOrders > 1");
			receiverConnection.start();

			TextMessage dummyMessage = senderSession.createTextMessage();
			dummyMessage.setStringProperty("NumberOfOrders", "2");
			dummyMessage.setText("testIdentifierConversion:1");
			sender.send(dummyMessage);

			TextMessage message = senderSession.createTextMessage();
			message.setIntProperty("NumberOfOrders", 2);
			message.setText("testIdentifierConversion:2");
			sender.send(message);

			TextMessage msg = (TextMessage) receiver.receive(ProviderLoader.TIMEOUT);
			assertTrue("No message was received", msg != null);
			if (msg == null)
				return;
			assertEquals("testIdentifierConversion:2", msg.getText());
		} catch (JMSException e) {
			fail(e.getMessage());
		}
	}

	/**
	 * Test the message selector using the filter example provided by the JMS specifications. <br />
	 * <ul>
	 * <li><code>"JMSType = 'car' AND color = 'blue' AND weight > 2500"</code></li>
	 * </ul>
	 */
	@Test
	public void testSelectorExampleFromSpecs() {
		try {
			//need close previous receiver for this session
			receiver.close();
			receiverConnection.stop();

			receiver = receiverSession.createReceiver(receiverQueue, "JMSType = 'car' AND color = 'blue' AND weight > 2500");
			receiverConnection.start();

			TextMessage dummyMessage = senderSession.createTextMessage();
			dummyMessage.setJMSType("car");
			dummyMessage.setStringProperty("color", "red");
			dummyMessage.setLongProperty("weight", 3000);
			dummyMessage.setText("testSelectorExampleFromSpecs:1");
			sender.send(dummyMessage);

			TextMessage message = senderSession.createTextMessage();
			message.setJMSType("car");
			message.setStringProperty("color", "blue");
			message.setLongProperty("weight", 3000);
			message.setText("testSelectorExampleFromSpecs:2");
			sender.send(message);

			TextMessage msg = (TextMessage) receiver.receive(ProviderLoader.TIMEOUT);
			assertTrue("No message was received", msg != null);
			if (msg == null)
				return;
			assertEquals("testSelectorExampleFromSpecs:2", msg.getText());
		} catch (JMSException e) {
			fail(e.getMessage());
		}
	}

	/**
	 * Test the ">" condition in message selector. <br />
	 * <ul>
	 * <li><code>"weight > 2500"</code> is <code>true</code> for 3000 and <code>false</code> for 1000</li>
	 * </ul>
	 */
	@Test
	public void testGreaterThan() {
		try {
			//need close previous receiver for this session
			receiver.close();
			receiverConnection.stop();

			receiver = receiverSession.createReceiver(receiverQueue, "weight > 2500");
			receiverConnection.start();

			TextMessage dummyMessage = senderSession.createTextMessage();
			dummyMessage.setLongProperty("weight", 1000);
			dummyMessage.setText("testGreaterThan:1");
			sender.send(dummyMessage);

			TextMessage message = senderSession.createTextMessage();
			message.setLongProperty("weight", 3000);
			message.setText("testGreaterThan:2");
			sender.send(message);

			TextMessage msg = (TextMessage) receiver.receive(ProviderLoader.TIMEOUT);
			assertTrue("No message was received", msg != null);
			if (msg == null)
				return;
			assertEquals("testGreaterThan:2", msg.getText());
		} catch (JMSException e) {
			fail(e.getMessage());
		}
	}

	/**
	 * Test the "=" condition in message selector. <br />
	 * <ul>
	 * <li><code>"weight > 2500"</code> is <code>true</code> for 2500 and <code>false</code> for 1000</li>
	 * </ul>
	 */
	@Test
	public void testEquals() {
		try {
			//need close previous receiver for this session
			receiver.close();
			receiverConnection.stop();

			receiver = receiverSession.createReceiver(receiverQueue, "weight = 2500");
			receiverConnection.start();

			TextMessage dummyMessage = senderSession.createTextMessage();
			dummyMessage.setLongProperty("weight", 1000);
			dummyMessage.setText("testEquals:1");
			sender.send(dummyMessage);

			TextMessage message = senderSession.createTextMessage();
			message.setLongProperty("weight", 2500);
			message.setText("testEquals:2");
			sender.send(message);

			TextMessage msg = (TextMessage) receiver.receive(ProviderLoader.TIMEOUT);
			assertTrue("No message was received", msg != null);
			if (msg == null)
				return;
			assertEquals("testEquals:2", msg.getText());
		} catch (JMSException e) {
			fail(e.getMessage());
		}
	}

	/**
	 * Test the "<>" (not equal) condition in message selector. <br />
	 * <ul>
	 * <li><code>"weight <> 2500"</code> is <code>true</code> for 1000 and <code>false</code> for 2500</li>
	 * </ul>
	 */
	@Test
	public void testNotEquals() {
		try {
			//need close previous receiver for this session
			receiver.close();
			receiverConnection.stop();

			receiver = receiverSession.createReceiver(receiverQueue, "weight <> 2500");
			receiverConnection.start();

			TextMessage dummyMessage = senderSession.createTextMessage();
			dummyMessage.setLongProperty("weight", 2500);
			dummyMessage.setText("testEquals:1");
			sender.send(dummyMessage);

			TextMessage message = senderSession.createTextMessage();
			message.setLongProperty("weight", 1000);
			message.setText("testEquals:2");
			sender.send(message);

			TextMessage msg = (TextMessage) receiver.receive(ProviderLoader.TIMEOUT);
			assertTrue("No message was received", msg != null);
			if (msg == null)
				return;
			assertEquals("testEquals:2", msg.getText());
		} catch (JMSException e) {
			fail(e.getMessage());
		}
	}

	/**
	 * Test the BETWEEN condition in message selector. <br />
	 * <ul>
	 * <li>"age BETWEEN 15 and 19" is <code>true</code> for 17 and <code>false</code> for 20</li>
	 * </ul>
	 */
	@Test
	public void testBetween() {
		try {
			//need close previous receiver for this session
			receiver.close();
			receiverConnection.stop();

			receiver = receiverSession.createReceiver(receiverQueue, "age BETWEEN 15 and 19");
			receiverConnection.start();

			TextMessage dummyMessage = senderSession.createTextMessage();
			dummyMessage.setIntProperty("age", 20);
			dummyMessage.setText("testBetween:1");
			sender.send(dummyMessage);

			TextMessage message = senderSession.createTextMessage();
			message.setIntProperty("age", 17);
			message.setText("testBetween:2");
			sender.send(message);

			TextMessage msg = (TextMessage) receiver.receive(ProviderLoader.TIMEOUT);
			assertTrue("Message not received", msg != null);
			if (msg == null)
				return;
			assertTrue("Message of another test: " + msg.getText(), msg.getText().startsWith("testBetween"));
			assertEquals("testBetween:2", msg.getText());

		} catch (JMSException e) {
			fail(e.getMessage());
		}
	}

	/**
	 * Test the IN condition in message selector. <br />
	 * <ul>
	 * <li>"Country IN ('UK', 'US', 'France')" is <code>true</code> for 'UK' and <code>false</code> for 'Peru'</li>
	 * </ul>
	 */
	@Test
	public void testIn() {
		try {
			//need close previous receiver for this session
			receiver.close();
			receiverConnection.stop();

			receiver = receiverSession.createReceiver(receiverQueue, "Country IN ('UK', 'US', 'France')");
			receiverConnection.start();

			TextMessage dummyMessage = senderSession.createTextMessage();
			dummyMessage.setStringProperty("Country", "Peru");
			dummyMessage.setText("testIn:1");
			sender.send(dummyMessage);

			TextMessage message = senderSession.createTextMessage();
			message.setStringProperty("Country", "UK");
			message.setText("testIn:2");
			sender.send(message);

			TextMessage msg = (TextMessage) receiver.receive(ProviderLoader.TIMEOUT);
			assertTrue("Message not received", msg != null);
			if (msg == null)
				return;
			assertTrue("Message of another test: " + msg.getText(), msg.getText().startsWith("testIn"));
			assertEquals("testIn:2", msg.getText());

		} catch (JMSException e) {
			fail(e.getMessage());
		}
	}

	/**
	 * Test the LIKE ... ESCAPE condition in message selector <br />
	 * <ul>
	 * <li>"underscored LIKE '\_%' ESCAPE '\'" is <code>true</code> for '_foo' and <code>false</code> for 'bar'</li>
	 * </ul>
	 */
	@Test
	public void testLikeEscape() {
		try {
			//need close previous receiver for this session
			receiver.close();
			receiverConnection.stop();

			receiver = receiverSession.createReceiver(receiverQueue, "underscored LIKE '\\_%' ESCAPE '\\'");
			receiverConnection.start();

			TextMessage dummyMessage = senderSession.createTextMessage();
			dummyMessage.setStringProperty("underscored", "bar");
			dummyMessage.setText("testLikeEscape:1");
			sender.send(dummyMessage);

			TextMessage message = senderSession.createTextMessage();
			message.setStringProperty("underscored", "_foo");
			message.setText("testLikeEscape:2");
			sender.send(message);

			TextMessage msg = (TextMessage) receiver.receive(ProviderLoader.TIMEOUT);
			assertTrue("Message not received", msg != null);
			if (msg == null)
				return;
			assertTrue("Message of another test: " + msg.getText(), msg.getText().startsWith("testLikeEscape"));
			assertEquals("testLikeEscape:2", msg.getText());

		} catch (JMSException e) {
			fail(e.getMessage());
		}
	}

	/**
	 * Test the LIKE condition with '_' in the pattern. <br />
	 * <ul>
	 * <li>"word LIKE 'l_se'" is <code>true</code> for 'lose' and <code>false</code> for 'loose'</li>
	 * </ul>
	 */
	@Test
	public void testLike_2() {
		try {
			//need close previous receiver for this session
			receiver.close();
			receiverConnection.stop();

			receiver = receiverSession.createReceiver(receiverQueue, "word LIKE 'l_se'");
			receiverConnection.start();

			TextMessage dummyMessage = senderSession.createTextMessage();
			dummyMessage.setStringProperty("word", "loose");
			dummyMessage.setText("testLike_2:1");
			sender.send(dummyMessage);

			TextMessage message = senderSession.createTextMessage();
			message.setStringProperty("word", "lose");
			message.setText("testLike_2:2");
			sender.send(message);

			TextMessage msg = (TextMessage) receiver.receive(ProviderLoader.TIMEOUT);
			assertTrue("Message not received", msg != null);
			if (msg == null)
				return;
			assertTrue("Message of another test: " + msg.getText(), msg.getText().startsWith("testLike_2"));
			assertEquals("testLike_2:2", msg.getText());

		} catch (JMSException e) {
			fail(e.getMessage());
		}
	}

	/**
	 * Test the LIKE condition with '%' in the pattern. <br />
	 * <ul>
	 * <li>"phone LIKE '12%3'" is <code>true</code> for '12993' and <code>false</code> for '1234'</li>
	 * </ul>
	 */
	@Test
	public void testLike_1() {
		try {
			//need close previous receiver for this session
			receiver.close();
			receiverConnection.stop();

			receiver = receiverSession.createReceiver(receiverQueue, "phone LIKE '12%3'");
			receiverConnection.start();

			TextMessage dummyMessage = senderSession.createTextMessage();
			dummyMessage.setStringProperty("phone", "1234");
			dummyMessage.setText("testLike_1:1");
			sender.send(dummyMessage);

			TextMessage message = senderSession.createTextMessage();
			message.setStringProperty("phone", "12993");
			message.setText("testLike_1:2");
			sender.send(message);

			TextMessage msg = (TextMessage) receiver.receive(ProviderLoader.TIMEOUT);
			assertTrue("Message not received", msg != null);
			if (msg == null)
				return;
			assertTrue("Message of another test: " + msg.getText(), msg.getText().startsWith("testLike_1"));
			assertEquals("testLike_1:2", msg.getText());

		} catch (JMSException e) {
			fail(e.getMessage());
		}
	}

	@Test
	/**
	 * Test the <code>NULL</code> value in message selector. <br />
	 * <ul>
	 * <li><code>"prop IS NULL"</code></li>
	 * </ul>
	 */
	public void testNull() {
		try {
			//need close previous receiver for this session
			receiver.close();
			receiverConnection.stop();

			receiver = receiverSession.createReceiver(receiverQueue, "prop_name IS NULL");
			receiverConnection.start();

			TextMessage dummyMessage = senderSession.createTextMessage();
			dummyMessage.setStringProperty("prop_name", "not null");
			dummyMessage.setText("testNull:1");
			sender.send(dummyMessage);

			TextMessage message = senderSession.createTextMessage();
			message.setText("testNull:2");
			sender.send(message);

			TextMessage msg = (TextMessage) receiver.receive(ProviderLoader.TIMEOUT);
			assertTrue(msg != null);
			if (msg == null)
				return;
			assertEquals("testNull:2", msg.getText());
		} catch (JMSException e) {
			fail(e.getMessage());
		}
	}
}

File: tests/brokertest/SelectorTest.cpp
/*
 * Copyright 2014-present IVK JSC. All Rights Reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

#include "SelectorTest.h"
#include <cstdint>
#include "cms/Message.h"
#include <cms/InvalidSelectorException.h>
#include <fake_cpp14.h>

typedef int8_t byte;

using namespace cms;

////////////////////////////////////////////////////////////////////////////////
void SelectorTest::SetUp() {
  cmsProvider = std::make_unique<CMSProvider>(getBrokerURL());
  cmsProvider->cleanUpDestination();
}

////////////////////////////////////////////////////////////////////////////////
TEST_F(SelectorTest, testBasicSelectors) {
  assertSelector("name = 'John'", false);
  assertSelector("name = 'James'", true);
  assertSelector("rank = 123", true);
  assertSelector("rank = 124", false);
  assertSelector("rank > 100", true);
  assertSelector("rank < 124", true);
  assertSelector("rank > 124", false);
  assertSelector("rank < 123", false);
  assertSelector("rank >= 123", true);
  assertSelector("rank <= 124", true);
  assertSelector("rank >= 124", false);
}

////////////////////////////////////////////////////////////////////////////////
TEST_F(SelectorTest, testBooleanSelector) {
  assertSelector("(trueProp OR falseProp) AND trueProp", true);
  assertSelector("(trueProp OR falseProp) AND falseProp", false);
  assertSelector("trueProp", true);
}

////////////////////////////////////////////////////////////////////////////////
TEST_F(SelectorTest, testJMSPropertySelectors) {
  assertSelector("JMSType = 'selector-test'", true);
  assertSelector("JMSType = 'crap'", false);
  assertSelector("JMSType = 'selector-test' OR JMSType='crap'", true);
}

////////////////////////////////////////////////////////////////////////////////
TEST_F(SelectorTest, testPropertyTypes) {
  assertSelector("byteProp = 123", true);
  assertSelector("byteProp = 10", false);

  assertSelector("byteProp2 = 33", true);
  assertSelector("byteProp2 = 10", false);

  assertSelector("shortProp = 123", true);
  assertSelector("shortProp = 10", false);

  assertSelector("intProp = 123", true);
  assertSelector("intProp = 10", false);

  assertSelector("longProp = 123", true);
  assertSelector("longProp = 10", false);

  assertSelector("floatProp = 123.0", true);
  assertSelector("floatProp = 10.0", false);

  assertSelector("doubleProp = 123.0", true);
  assertSelector("doubleProp = 10.0", false);
}

////////////////////////////////////////////////////////////////////////////////
TEST_F(SelectorTest, testAndSelectors) {
  assertSelector("name = 'James' and rank < 200", true);
  assertSelector("name = 'James' and rank > 200", false);
  assertSelector("name = 'Foo' and rank < 200", false);
  assertSelector("unknown = 'Foo' and anotherUnknown < 200", false);
}

////////////////////////////////////////////////////////////////////////////////
TEST_F(SelectorTest, testOrSelectors) {
  assertSelector("name = 'James' or rank < 200", true);
  assertSelector("name = 'James' or rank > 200", true);
  assertSelector("name = 'Foo' or rank < 200", true);
  assertSelector("name = 'Foo' or rank > 200", false);
  assertSelector("unknown = 'Foo' or anotherUnknown < 200", false);
}

////////////////////////////////////////////////////////////////////////////////
TEST_F(SelectorTest, testBetween) {
  assertSelector("rank between 100 and 150", true);
  assertSelector("rank between 10 and 120", false);
}

////////////////////////////////////////////////////////////////////////////////
TEST_F(SelectorTest, testIn) {
  assertSelector("name in ('James', 'Bob', 'Gromit')", true);
  assertSelector("name in ('Bob', 'James', 'Gromit')", true);
  assertSelector("name in ('Gromit', 'Bob', 'James')", true);

  assertSelector("name in ('Gromit', 'Bob', 'Cheddar')", false);
  assertSelector("name not in ('Gromit', 'Bob', 'Cheddar')", true);
}

////////////////////////////////////////////////////////////////////////////////
TEST_F(SelectorTest, testIsNull) {
  assertSelector("dummy is null", true);
  assertSelector("dummy is not null", false);
  assertSelector("name is not null", true);
  assertSelector("name is null", false);
}

////////////////////////////////////////////////////////////////////////////////
TEST_F(SelectorTest, testStringQuoteParsing) { assertSelector("quote = '''In God We Trust'''", true); }

////////////////////////////////////////////////////////////////////////////////
TEST_F(SelectorTest, testLikeComparisons) {
  assertSelector("quote LIKE '''In G_d We Trust'''", true);
  assertSelector("quote LIKE '''In Gd_ We Trust'''", false);
  assertSelector("quote NOT LIKE '''In G_d We Trust'''", false);
  assertSelector("quote NOT LIKE '''In Gd_ We Trust'''", true);

  assertSelector("foo LIKE '%oo'", true);
  assertSelector("foo LIKE '%ar'", false);
  assertSelector("foo NOT LIKE '%oo'", false);
  assertSelector("foo NOT LIKE '%ar'", true);

  assertSelector("foo LIKE '!_%' ESCAPE '!'", true);
  assertSelector("quote LIKE '!_%' ESCAPE '!'", false);
  assertSelector("foo NOT LIKE '!_%' ESCAPE '!'", false);
  assertSelector("quote NOT LIKE '!_%' ESCAPE '!'", true);

  assertSelector("punctuation LIKE '!#$&()*+,-./:;<=>?@[\\]^`{|}~'", true);
}

TEST_F(SelectorTest, testInvalidSelector) { assertInvalidSelector("=TEST 'test'"); }

////////////////////////////////////////////////////////////////////////////////
void SelectorTest::assertInvalidSelector(const std::string &selector) {
  std::unique_ptr<cms::Session> session(cmsProvider->getConnection()->createSession());

  EXPECT_THROW(std::unique_ptr<cms::MessageConsumer> consumer(session->createConsumer(cmsProvider->getDestination(), selector)),
               cms::InvalidSelectorException)
      << "should throw InvalidSelectorException";
}

////////////////////////////////////////////////////////////////////////////////
void SelectorTest::assertSelector(const std::string &selector, bool expected) {
  std::unique_ptr<cms::Session> session(cmsProvider->getConnection()->createSession());
  std::unique_ptr<cms::MessageProducer> producer(session->createProducer(cmsProvider->getDestination()));
  std::unique_ptr<cms::MessageConsumer> consumer(session->createConsumer(cmsProvider->getDestination(), selector));

  std::unique_ptr<cms::Message> msg(session->createMessage());

  doMessage(msg.get());
  producer->send(msg.get());

  const std::unique_ptr<cms::Message> msgret(consumer->receive(3000));

  if (expected) {
    EXPECT_TRUE(msgret != nullptr);
  } else {
    doClean();
    EXPECT_TRUE(msgret == nullptr);
  }

  consumer->close();
  producer->close();
  session->close();
}

////////////////////////////////////////////////////////////////////////////////
void SelectorTest::doMessage(cms::Message *message) {
  message->setCMSType("selector-test");
  message->setStringProperty("name", "James");
  message->setStringProperty("location", "London");

  message->setByteProperty("byteProp", static_cast<byte>(123));
  message->setByteProperty("byteProp2", static_cast<byte>(33));
  message->setShortProperty("shortProp", static_cast<short>(123));
  message->setIntProperty("intProp", static_cast<int>(123));
  message->setLongProperty("longProp", static_cast<long>(123));
  message->setFloatProperty("floatProp", static_cast<float>(123.0f));
  message->setDoubleProperty("doubleProp", static_cast<double>(123.0));

  message->setIntProperty("rank", 123);
  message->setIntProperty("version", 2);
  message->setStringProperty("quote", "'In God We Trust'");
  message->setStringProperty("foo", "_foo");
  message->setStringProperty("punctuation", "!#$&()*+,-./:;<=>?@[\\]^`{|}~");
  message->setBooleanProperty("trueProp", true);
  message->setBooleanProperty("falseProp", false);
}

////////////////////////////////////////////////////////////////////////////////
void SelectorTest::doClean() {
  cms::Session *session(cmsProvider->getSession());
  (void)session;
  cms::MessageConsumer *consumer(cmsProvider->getConsumer());
  consumer->start();
  // lets consume any outstanding messages from previous test runs
  cms::Message *message;
  while ((message = consumer->receive(3000)) != nullptr) {
    delete message;
  }
  consumer->stop();
}
////////////////////////////////////////////////////////////////////////////////
void SelectorTest::TearDown() {}
